# День 00 - Сбор данных

In [84]:
import os
import random

import numpy as np
import pandas as pd

In [2]:
VERSION = "v2"

In [3]:
random.seed(42)

In [4]:
# Reduce memory usage
def reduce_mem_usage(df, verbose=True):
    numerics = ['int16', 'int32', 'int64', 'float16', 'float32', 'float64']
    start_mem = df.memory_usage(deep=True).sum() / 1024**2    
    for col in df.columns:
        col_type = df[col].dtypes
        if col_type in numerics:
            c_min = df[col].min()
            c_max = df[col].max()
            if str(col_type)[:3] == 'int':  # for integers
                if c_min > np.iinfo(np.int8).min and c_max < np.iinfo(np.int8).max:
                    df[col] = df[col].astype(np.int8)
                elif c_min > np.iinfo(np.int16).min and c_max < np.iinfo(np.int16).max:
                    df[col] = df[col].astype(np.int16)
                elif c_min > np.iinfo(np.int32).min and c_max < np.iinfo(np.int32).max:
                    df[col] = df[col].astype(np.int32)
                elif c_min > np.iinfo(np.int64).min and c_max < np.iinfo(np.int64).max:
                    df[col] = df[col].astype(np.int64)  
            else:  # for floats.
                if c_min > np.finfo(np.float16).min and c_max < np.finfo(np.float16).max:
                    df[col] = df[col].astype(np.float16)
                elif c_min > np.finfo(np.float32).min and c_max < np.finfo(np.float32).max:
                    df[col] = df[col].astype(np.float32)
                else:
                    df[col] = df[col].astype(np.float64)    
    end_mem = df.memory_usage(deep=True).sum() / 1024**2
    if verbose: print('Mem. usage decreased to {:5.2f} Mb ({:.1f}% reduction)'.format(end_mem, 100 * (start_mem - end_mem) / start_mem))
    return df

# Загрузка данных

In [5]:
train = pd.read_csv('../data/raw/final_train.csv', index_col=0)

In [6]:
train.head(10)

load_dttm_dm              hflat_1m    hlid_1m     hgid_1m  \
0  2022-05-13 12:54:00.235  11726022852127600908  579093646   9107155.0   
1  2022-05-13 12:54:00.235   3368930564640238896  370185602  29928155.0   
2  2022-05-13 12:54:00.235   1093840297643692259  388961722  13089550.0   
3  2022-05-13 12:54:00.235  14270383511494010742  367636608  18322138.0   
4  2022-05-13 12:54:00.235    918694842307981215  409108209  15069579.0   
5  2022-05-13 12:54:00.235   2580274487661669249  370131132  12585784.0   
6  2022-05-13 12:54:00.235    677166249813083950  432647573  26458375.0   
7  2022-05-13 12:54:00.235     15486421759062996  388943532  11051427.0   
8  2022-05-13 12:54:00.235   6580846247176123020  388943532  11051427.0   
9  2022-05-13 12:54:00.235   4574582212156368972  370164324   4912701.0   

   lifetime_ota_1m  lifetime_shpd_1m  lifetime_iptv_1m  lifetime_tp_ota_1m  \
0          18597.0           18597.0            3031.0             18597.0   
1              NaN           18597.0           18597.0                 NaN   
2              NaN           18597.0             366.0                 NaN   
3          18597.0             794.0           18597.0             18597.0   
4          18597.0           18597.0               NaN             18597.0   
5          18597.0           18597.0               NaN             18597.0   
6          18597.0           18597.0           18597.0             18597.0   
7              NaN           18597.0            6519.0                 NaN   
8          18597.0           18597.0           18597.0             18597.0   
9              NaN             193.0               NaN                 NaN   

   lifetime_tp_shpd_1m  lifetime_tp_iptv_1m  ...  view_time_film_fantasy_2m  \
0              18597.0              18597.0  ...                        NaN   
1              18597.0              18597.0  ...                        NaN   
2              18597.0              18597.0  ...                        NaN   
3                673.0              18597.0  ...                        NaN   
4              18597.0                  NaN  ...                        NaN   
5              18597.0                  NaN  ...                        NaN   
6              18597.0              18597.0  ...                        NaN   
7              18597.0               6519.0  ...                        NaN   
8              18597.0              18597.0  ...                        NaN   
9              18597.0                  NaN  ...                        NaN   

   view_time_film_fantasy_3m  view_time_film_fantasy_p1m  \
0                        NaN                         NaN   
1                        NaN                         NaN   
2                        NaN                         NaN   
3                        NaN                         NaN   
4                        NaN                         NaN   
5                        NaN                         NaN   
6                        NaN                         NaN   
7                        NaN                         NaN   
8                        NaN                         NaN   
9                        NaN                         NaN   

   view_time_film_fantasy_p2m  view_time_film_fantasy_p3m  \
0                         NaN                         NaN   
1                         NaN                         NaN   
2                         NaN                         NaN   
3                         NaN                         NaN   
4                         NaN                         NaN   
5                         NaN                         NaN   
6                         NaN                         NaN   
7                         NaN                         NaN   
8                         NaN                         NaN   
9                         NaN                         NaN   

   view_time_film_fantasy_d1m  view_time_film_fantasy_d2m  \
0                         NaN                         NaN   
1   

In [7]:
train.info()

<class 'pandas.core.frame.DataFrame'>
Index: 160461 entries, 0 to 160460
Columns: 4369 entries, load_dttm_dm to client_id
dtypes: float64(4340), int64(19), object(9), uint64(1)
memory usage: 5.2+ GB


In [8]:
val = pd.read_csv('../data/raw/final_valid.csv', index_col=0)

In [9]:
val.info()

<class 'pandas.core.frame.DataFrame'>
Index: 173304 entries, 0 to 173303
Columns: 4369 entries, load_dttm_dm to client_id
dtypes: float64(4314), int64(45), object(9), uint64(1)
memory usage: 5.6+ GB


Нет данных по оттоку в тесте. **test не используем** 

In [10]:
# test = pd.read_csv('../data/raw/final_test.csv', index_col=0)

In [11]:
# test.info()

In [12]:
# "label" in test.columns

In [13]:
attr = pd.read_excel("../data/raw/attributes.xlsx")

In [14]:
attr

Атрибут                                 Предметная область  \
0                  nls                                                NaN   
1              mrf_key                                                NaN   
2           period_key                                                NaN   
3              san_cnt  Агрегированные данные по IPTV (покупки, телесм...   
4              mac_cnt  Агрегированные данные по IPTV (покупки, телесм...   
...                ...                                                ...   
1015           prob100                       Посещение сайтов конкурентов   
1016  min_long_days100                       Посещение сайтов конкурентов   
1017        total_prob                       Посещение сайтов конкурентов   
1018     weighted_prob                       Посещение сайтов конкурентов   
1019    weighted_prob2                       Посещение сайтов конкурентов   

                                               Описание  
0                                          лицевой счет  
1                                                id МРФ  
2                                       отчетный период  
3                             Количество уникальных SAN  
4                             Количество уникальных MAC  
...                                                 ...  
1015  Количество дней с посещением сайтов конкуренто...  
1016  Минимальный период от посещения сайта конкурен...  
1017    Количество дней с посещением сайтов конкурентов  
1018  Взвешанная вероятность посещения сайтов конкур...  
1019  Взвешанная вероятность посещения сайтов конкур...  

[1020 rows x 3 columns]

In [15]:
suffix = pd.read_table("../data/raw/1m, 2m etc.txt", sep="\t", encoding="cp1251")

In [16]:
suffix

1m – сумма за текущий период
0             2m – сумма за последние 2 периода
1             3m – сумма за последние 3 периода
2                 p1m – сумма за прошлый период
3   p2m – сумма с 4-го по 3-ий период в прошлом
4  p3m – сумма за 6-ий по 4-ый период в прошлом
5                        d1m – отношение 1m/p1m
6                        d2m – отношение 2m/p2m
7                        d3m – отношение 3m/p3m

# Small EDA

In [17]:
columns_features = pd.Series(train.columns)

In [18]:
columns_features

0                     load_dttm_dm
1                         hflat_1m
2                          hlid_1m
3                          hgid_1m
4                  lifetime_ota_1m
                   ...            
4364    view_time_film_fantasy_d1m
4365    view_time_film_fantasy_d2m
4366    view_time_film_fantasy_d3m
4367                         label
4368                     client_id
Length: 4369, dtype: object

In [19]:
didnt_match_features = columns_features[~columns_features.str.startswith(tuple(attr["Атрибут"]))]

In [20]:
didnt_match_features.str.split('_').apply(lambda x: x[0]).unique()

array(['load', 'asr', 'appm', 'view', 'label', 'client'], dtype=object)

In [21]:
didnt_match_features_v2 = didnt_match_features[~didnt_match_features.str.startswith(("asr", "appm", "view"))]

In [22]:
didnt_match_features_v2

0       load_dttm_dm
4367           label
4368       client_id
dtype: object

Есть описания для всех признаков кроме сегментов _"asr", "appm", "view"_. Они **не попадут в признаки**

# Разделение признаков

In [23]:
using_features = pd.read_excel("../data/raw/attribute.xlsx")

In [24]:
using_features["Предметная область"].value_counts()

Предметная область
Сбои                                                      535
Обращения клиентов                                        174
Услуги у клиента                                          170
Общая информация о клиенте                                 67
Агрегированные данные по IPTV (покупки, телесмотрение)     41
Посещение сайтов конкурентов                               25
География                                                   7
Name: count, dtype: int64

## Client

In [25]:
# Remove Some Prefix

In [26]:
client_columns = using_features[using_features["Предметная область"] == "Общая информация о клиенте"]["Атрибут"]

In [27]:
len(client_columns)

67

In [28]:
client_columns_original = list(columns_features[columns_features.str.startswith(tuple(client_columns))].values)
client_columns_original.extend(["client_id"])
client_columns_original.extend(client_columns.values)

In [29]:
remove_columns = set(
    ['lifetime_iptv_1m', 'lifetime_ota_1m', 'lifetime_shpd_1m', 
    'lifetime_tp_iptv_1m', 'lifetime_tp_ota_1m', 'lifetime_tp_shpd_1m']
)

In [30]:
client_columns_original = list(set(client_columns_original).difference(remove_columns))

In [31]:
client_columns_original = list(set(columns_features) & set(client_columns_original))

In [32]:
len(client_columns_original)

283

## Geography

In [33]:
geography_columns = using_features[using_features["Предметная область"] == "География"]["Атрибут"]

In [34]:
len(geography_columns)

7

In [35]:
geography_columns_original = list(columns_features[columns_features.str.startswith(tuple(geography_columns))].values)
geography_columns_original.extend(geography_columns.values)

In [36]:
geography_columns_original = list(set(columns_features) & set(geography_columns_original))

In [37]:
len(geography_columns_original)

7

## Competitors

In [38]:
competitors_columns = using_features[using_features["Предметная область"] == "Посещение сайтов конкурентов"]["Атрибут"]

In [39]:
len(competitors_columns)

25

In [40]:
competitors_columns_original = list(columns_features[columns_features.str.startswith(tuple(competitors_columns))].values)
competitors_columns_original.extend(competitors_columns)

In [41]:
competitors_columns_original = list(set(columns_features) & set(competitors_columns_original))

In [42]:
len(competitors_columns_original)

0

Посещение сайтов конкурентов - нет этих признаков

## Services

In [43]:
services_columns = using_features[using_features["Предметная область"] == "Услуги у клиента"]["Атрибут"]

In [44]:
len(services_columns)

170

In [45]:
services_columns_original = list(columns_features[columns_features.str.startswith(tuple(services_columns))].values)
services_columns_original.extend(services_columns)

In [46]:
services_columns_original = list(set(columns_features) & set(services_columns_original))

In [47]:
len(services_columns_original)

1414

## Appeals

In [48]:
appeals_columns = using_features[using_features["Предметная область"] == "Обращения клиентов"]["Атрибут"]

In [49]:
len(appeals_columns)

174

In [50]:
appeals_columns_original = list(columns_features[columns_features.str.startswith(tuple(appeals_columns))].values)
appeals_columns_original.extend(appeals_columns)

In [51]:
appeals_columns_original = list(set(columns_features) & set(appeals_columns_original))

In [52]:
len(appeals_columns_original)

1566

## Failures

In [53]:
failures_columns = using_features[using_features["Предметная область"] == "Сбои"]["Атрибут"]

In [54]:
len(failures_columns)

535

In [55]:
failures_columns_original = list(columns_features[columns_features.str.startswith(tuple(failures_columns))].values)
failures_columns_original.extend(failures_columns)

In [56]:
failures_columns_original = list(set(columns_features) & set(failures_columns_original))

In [57]:
len(failures_columns_original)

9

## IPTV

In [58]:
iptv_columns = using_features[using_features["Предметная область"] == "Агрегированные данные по IPTV (покупки, телесмотрение)"]["Атрибут"]

In [59]:
len(iptv_columns)

41

In [60]:
iptv_columns_original = list(columns_features[columns_features.str.startswith(tuple(iptv_columns))].values)
iptv_columns_original.extend(iptv_columns)

In [61]:
iptv_columns_original = list(set(columns_features) & set(iptv_columns_original))

In [62]:
len(iptv_columns_original)

369

## Merge Assert

In [63]:
used_columns = client_columns_original \
                + geography_columns_original \
                + competitors_columns_original \
                + services_columns_original \
                + appeals_columns_original \
                + failures_columns_original \
                + iptv_columns_original

In [64]:
len(set(used_columns))

3639

In [65]:
len(set(columns_features))

4369

In [66]:
len(set(columns_features.values.tolist()) -  set(used_columns)) 

730

Все сходится. Все эти 730 признаков без аннотации в attributes

# Сохраняем Train и Test (a.k.a Validation) и Ресемплинг

In [67]:
SAVE_PATH = f"../data/{VERSION}/"

In [68]:
train = train.groupby('label', group_keys=False).apply(lambda x: x.sample(frac=0.2))
val = val.groupby('label', group_keys=False).apply(lambda x: x.sample(frac=0.2))

/var/folders/qv/19_404_959jdcbdt74fg3hbm0000gn/T/ipykernel_14192/351966677.py:1: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  train = train.groupby('label', group_keys=False).apply(lambda x: x.sample(frac=0.2))
/var/folders/qv/19_404_959jdcbdt74fg3hbm0000gn/T/ipykernel_14192/351966677.py:2: DeprecationWarning: DataFrameGroupBy.apply operated on the grouping columns. This behavior is deprecated, and in a future version of pandas the grouping columns will be excluded from the operation. Either pass `include_groups=False` to exclude the groupings or explicitly select the grouping columns after groupby to silence this warning.
  val = val.groupby('label', group_keys=False).apply(lambda x: x.sample(f

In [69]:
train.shape

(32092, 4369)

In [70]:
val.shape

(34661, 4369)

In [71]:
train[used_columns].to_csv("../data/v2/day01/dataset_01_06.csv", index=False)
val[used_columns].to_csv("../data/v2/day01/dataset_07_12.csv", index=False)

In [72]:
train[used_columns].to_csv("../data/v2/day02/dataset_01_06.csv", index=False)
val[used_columns].to_csv("../data/v2/day02/dataset_07_12.csv", index=False)

# Сохраняем все сплитованные данные

Валидацию отставляем на **2ой день**

In [73]:
from sklearn.model_selection import train_test_split

In [74]:
train_01_03, train_04_06 = train_test_split(train, test_size=0.5, random_state=42)

## Client

In [75]:
client_columns_original.sort()

In [76]:
client_columns_original;

In [77]:
mapper = {
    'client_id': "id",
    "gender_1m": "gender",
}

In [78]:
data_client_2021_01_03 = train_01_03[client_columns_original].rename(columns=mapper)
data_client_2021_04_06 = train_04_06[client_columns_original].rename(columns=mapper)

In [82]:
data_client_2021_01_03.shape

(16046, 283)

In [83]:
data_client_2021_04_06.shape

(16046, 283)

In [85]:
data_client_2021_01_03.to_csv(os.path.join(SAVE_PATH, "day00/client_2021_01_03.csv"), index_label=False)
data_client_2021_04_06.to_csv(os.path.join(SAVE_PATH, "day00/client_2021_04_06.csv"), index_label=False)

## Geography

In [86]:
geography_columns_original.sort()

In [87]:
geography_columns_original;

In [88]:
geography_columns_original.extend(["client_id"])

In [89]:
geography_columns_original

['hex_id10_1m',
 'hex_id11_1m',
 'hex_id12_1m',
 'hex_id6_1m',
 'hex_id7_1m',
 'hex_id8_1m',
 'hex_id9_1m',
 'client_id']

In [90]:
data_geography_2021_01_03 = train_01_03[geography_columns_original]
data_geography_2021_04_06 = train_04_06[geography_columns_original]

In [91]:
data_geography_2021_01_03.to_csv(os.path.join(SAVE_PATH, "day00/geography_2021_01_03.csv"), index_label=False)
data_geography_2021_04_06.to_csv(os.path.join(SAVE_PATH, "day00/geography_2021_04_06.csv"), index_label=False)

## Services

In [92]:
services_columns_original.sort()

In [93]:
services_columns_original;

In [94]:
services_columns_original.extend(["client_id"])

In [95]:
data_services_2021_01_03 = train_01_03[services_columns_original]
data_services_2021_04_06 = train_04_06[services_columns_original]

In [96]:
data_services_2021_01_03.to_csv(os.path.join(SAVE_PATH, "day00/services_2021_01_03.csv"), index_label=False)
data_services_2021_04_06.to_csv(os.path.join(SAVE_PATH, "day00/services_2021_04_06.csv"), index_label=False)

## Appeals

In [97]:
appeals_columns_original.sort()

In [98]:
appeals_columns_original;

In [99]:
appeals_columns_original.extend(["client_id"])

In [100]:
data_appeals_2021_01_03 = train_01_03[appeals_columns_original]
data_appeals_2021_04_06 = train_04_06[appeals_columns_original]

In [101]:
data_appeals_2021_01_03.to_csv(os.path.join(SAVE_PATH, "day00/appeals_2021_01_03.csv"), index_label=False)
data_appeals_2021_04_06.to_csv(os.path.join(SAVE_PATH, "day00/appeals_2021_04_06.csv"), index_label=False)

## Failures

In [102]:
failures_columns_original.sort()

In [103]:
failures_columns_original;

In [104]:
failures_columns_original.extend(["client_id"])

In [105]:
data_failures_2021_01_03 = train_01_03[failures_columns_original]
data_failures_2021_04_06 = train_04_06[failures_columns_original]

In [106]:
data_failures_2021_01_03.to_csv(os.path.join(SAVE_PATH, "day00/failures_2021_01_03.csv"), index_label=False)
data_failures_2021_04_06.to_csv(os.path.join(SAVE_PATH, "day00/failures_2021_04_06.csv"), index_label=False)

## IPTV

In [107]:
iptv_columns_original.sort()

In [108]:
iptv_columns_original;

In [109]:
iptv_columns_original.extend(["client_id"])

In [110]:
data_iptv_2021_01_03 = train_01_03[iptv_columns_original]
data_iptv_2021_04_06 = train_04_06[iptv_columns_original]

In [111]:
data_iptv_2021_01_03.to_csv(os.path.join(SAVE_PATH, "day00/iptv_2021_01_03.csv"), index_label=False)
data_iptv_2021_04_06.to_csv(os.path.join(SAVE_PATH, "day00/iptv_2021_04_06.csv"), index_label=False)